### PR TITLE
Add PKCS11 KMSClient

### PR DESCRIPTION
**Pull request by solcates**
This PR adds a PKCS11 KMSClient to tink using AES256GCM for those that would like to leverage an HSM for their key management and key entropy.

It uses the [crypto11](https://github.com/ThalesIgnite/crypto11) library to expose simple Go crypto interfaces to a  PKCS11 device (HSM, Token, SoftHSM, etc...)

The keyURI introduced is pkcs11://KeyID in where KeyID is a unique ID in your device (ID only, labels currently not allowed).

The README.md goes into some details about the 2 modes of operation, as well as some basic benchmarks using both HSM and Wrapped keys as your tink.AEAD.

P.S. - I have struggled to get Bazelisk to build the project, so hopefully the bazel modifications work for someone with a working Bazel environment.

### CODE REVIEW

**Comment by rbroggi** [comment]
Nice one :+1:

**Comment by solcates** [comment]
For sure.  Happy to toss in a few other languages, but Go is my language of choice these days :)

File: go/integration/pkcs11kms/README.md [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+HSM Keys reside in the pkcs11 token/slot, and a cipher.AEAD handle to the PKCS11 key handle which is then used to support the tink.AEAD interface transparently.
+
+HSM Keys are:
+- SecretKeys in the HSM
+- Session based using a pkcs11 session to the HSM for each KMSClient provided
+- The AEAD handle will be accessible as long as the session is open
+- Good for "remove the HSM" use cases that will close all access to the AEAD even to tink
+
+The keyURI for a HSM key is `pkcs11://ee2c5de3-335c-428b-8168-5daf299e97f3`  where `ee2c5de3-335c-428b-8168-5daf299e97f3` is the key ID (not label) in the device to use.  To have the key auto created if it doesnt' exist yet, simply proved the "autogen" bool to the client.
```
Comment (Baha-sk) [inline go/integration/pkcs11kms/README.md]: change:
simply `proved` the "autogen" bool to the client. 
==>
simply `provide` the "autogen" bool to the client.

File: go/integration/pkcs11kms/key.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+	k = &Key{}
+	var u *url.URL
+	if u, err = url.Parse(keyURI); err != nil {
+		return
+	}
+	if k.id, err = uuid.Parse(u.Host); err != nil {
+		return
+	}
+
+	if q := u.Query(); q != nil {
```
Comment (Baha-sk) [inline go/integration/pkcs11kms/key.go]: if `q` is nil, then the function returns `nil, nil` 

this is not a good Go convention as one would expect either a non nil error or a non nil *Key as a return value

File: go/integration/pkcs11kms/key_test.go [hunk 3]
```diff
@@ -0,0 +1,147 @@
+package pkcs11kms
+
+import (
+	"context"
+	"fmt"
+	"github.com/ThalesIgnite/crypto11"
```
Comment (Baha-sk) [inline go/integration/pkcs11kms/key_test.go]: move non Go core libraries under core ones, ie add a space after line 13 and move them under there

File: go/integration/pkcs11kms/pkcs11_aead.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
+package pkcs11kms
+
+import (
+	"bytes"
+	"crypto/cipher"
+	"fmt"
+	"github.com/ThalesIgnite/crypto11"
+	"github.com/ThalesIgnite/gose"
+	"github.com/ThalesIgnite/gose/jose"
+	"github.com/google/tink/go/tink"
```
Comment (Baha-sk) [inline go/integration/pkcs11kms/pkcs11_aead.go]: same here.. non core libraries should go after core ones

File: go/integration/pkcs11kms/pkcs11_aead_test.go [hunk 5]
```diff
@@ -0,0 +1,223 @@
+package pkcs11kms
+
+import (
+	"fmt"
+	"github.com/ThalesIgnite/crypto11"
+	"github.com/ThalesIgnite/gose"
```
Comment (Baha-sk) [inline go/integration/pkcs11kms/pkcs11_aead_test.go]: idem

File: go/integration/pkcs11kms/pkcs11_client.go [hunk 6]
```diff
@@ -0,0 +1,46 @@
+package pkcs11kms
+
+import (
+	"context"
+	"github.com/ThalesIgnite/crypto11"
+	"github.com/google/tink/go/core/registry"
+	"github.com/google/tink/go/tink"
```
Comment (Baha-sk) [inline go/integration/pkcs11kms/pkcs11_client.go]: same here

File: go/integration/pkcs11kms/pkcs11_client_test.go [hunk 7]
```diff
@@ -0,0 +1,140 @@
+package pkcs11kms
+
+import (
+	"context"
+	"github.com/ThalesIgnite/crypto11"
```
Comment (Baha-sk) [inline go/integration/pkcs11kms/pkcs11_client_test.go]: and here

File: go/integration/pkcs11kms/README.md (hunk 1 above)
Comment (solcates) [reply go/integration/pkcs11kms/README.md]: good catch

File: go/integration/pkcs11kms/key.go (hunk 2 above)
Comment (solcates) [reply go/integration/pkcs11kms/key.go]: I'm not sure I follow. 

If `q` is nil, then the function will still return `*Key,nil` as k is  defined on line 17, nil as there is no "wrappedKey" for the wrapped key mode.

The expected outcome is a *Key with optional wrappedKey(from a query param), or an error.

File: go/integration/pkcs11kms/key_test.go (hunk 3 above)
Comment (solcates) [reply go/integration/pkcs11kms/key_test.go]: 👍

File: go/integration/pkcs11kms/pkcs11_client_test.go (hunk 7 above)
Comment (solcates) [reply go/integration/pkcs11kms/pkcs11_client_test.go]: 👍

File: go/integration/pkcs11kms/key.go (hunk 2 above)
Comment (Baha-sk) [reply go/integration/pkcs11kms/key.go]: I see, I should've read more closely the comment of the function saying: `optionally provide it's protected wrapped key`.

looks good then.

Comment (solcates) [reply go/integration/pkcs11kms/key.go]: no worries, just wanted to make sure i wasn't missing something :)

appreciate the review.

File: go/integration/pkcs11kms/key_test.go [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+func setupTests(t testing.TB) func(t testing.TB) {
+
+	testCtx = context.Background()
+	var err error
+
+	if os.Getenv("SOFTHSM_LIBRARY") != "" {
+		testHSMConfigured = true
+	}
+
+	//logrus.SetLevel(logrus.DebugLevel)
```
Comment (Baha-sk) [inline go/integration/pkcs11kms/key_test.go]: should probably remove `logrus` from Tink

Comment (Baha-sk) [reply go/integration/pkcs11kms/key_test.go]: and its references elsewhere..

Comment (solcates) [reply go/integration/pkcs11kms/key_test.go]: Agreed, forgot it was there during our smoke testing earlier.

File: go/WORKSPACE [hunk 9]
```diff
@@ -265,3 +281,42 @@ go_repository(
     importpath = "github.com/hashicorp/vault",
     tag = "v1.2.3",
 )
+go_repository(
+    name = "com_github_google_uuid",
+    importpath = "github.com/google/uuid",
+    tag = "v1.0.0",
+)
+go_repository(
```
Comment (Baha-sk) [inline go/WORKSPACE]: I think you missed this entry

File: go/go.sum [hunk 10]
```diff
@@ -24,23 +35,43 @@ github.com/google/go-cmp v0.3.0 h1:crn/baboCvb5fXaQ0IJ1SGTsTVrWpDsCWC8EGETZijY=
...
+github.com/konsorten/go-windows-terminal-sequences v1.0.1/go.mod h1:T0+1ngSBFLxvqU3pZ+m/2kptfBszLMUkC4ZK/EgS/cQ=
+github.com/miekg/pkcs11 v1.0.3-0.20190429190417-a667d056470f h1:eVB9ELsoq5ouItQBr5Tj334bhPJG/MX+m7rTchmzVUQ=
+github.com/miekg/pkcs11 v1.0.3-0.20190429190417-a667d056470f/go.mod h1:XsNlhZGX73bx86s2hdc/FuaLm2CPZJemRLMA+WTFxgs=
+github.com/pkg/errors v0.8.1 h1:iURUrRGxPUNPdy5/HRSm+Yj6okJ6UtLINN0Q9M4+h3I=
+github.com/pkg/errors v0.8.1/go.mod h1:bwawxfHBFNV+L2hUp1rHADufV3IMtnDRdf1r5NINEl0=
 github.com/pmezard/go-difflib v1.0.0 h1:4DBwDE0NGyQoBHbLQYPwSUPoCMWR5BEzIk/f1lZbAQM=
 github.com/pmezard/go-difflib v1.0.0/go.mod h1:iKH77koFhYxTK1pcRnkKkqfTogsbg7gZNVY4sRDYZ/4=
+github.com/rogpeppe/fastuuid v1.2.0/go.mod h1:jVj6XXZzXRy/MSR5jhDC/2q6DgLz+nrA6LYCDYWNEvQ=
+github.com/sirupsen/logrus v1.2.0 h1:juTguoYk5qI21pwyTXY3B3Y5cOTH3ZUyZCg1v/mihuo=
+github.com/sirupsen/logrus v1.2.0/go.mod h1:LxeOpSwHxABJmUn/MG1IvRgCAasNZTLOkJPxbbu5VWo=
```
Comment (Baha-sk) [inline go/go.sum]: and these two lines

File: go/WORKSPACE (hunk 9 above)
Comment (solcates) [reply go/WORKSPACE:289]: I did thank you

File: go/go.sum (hunk 10 above)
Comment (solcates) [reply go/go.sum]: Yep, thanks

**Comment by solcates** [comment]
Sorry for the delayed clean up, just got a chance to get back to this.

**Comment by juergw** [comment]
Sorry that nobody replied to this until now.

Adding this code to Tink's code base would mean that we commit to supporting, and we can't do that. We don't have the time for that. Also, we don't want to add additional dependencies to the Tink code base.

But maybe this integration could be added directly to the crypto11 library? Or, if they don't want a dependency on Tink, it could be made its own library.

**Pull request closed without merging**